File: ml/src/wikidetect/sweep.py
"""Category-wide triage sweep: score EVERY article in a WikiProject
AI-Cleanup category and write a ranked report.

    wikidetect sweep --category "Category:Articles containing suspected
        AI-generated texts from November 2025" [--limit N] [--out NAME]

Replaces the retired sweep.js. Same record schema and resume semantics
(append JSONL keyed by title; rerun skips what's done; the ranked .md is
rewritten at the end), with three changes:
  - fetching is concurrent (async, polite semaphore) and overlaps scoring
  - inference is batched with dynamic padding, through the shared cache
  - records carry the FULL text and revid, so a finding can be promoted
    into an eval case even after the live article is cleaned up
"""

import asyncio
import json

from . import config
from .cache import ScoreCache
from .detect import load_calibration, verdict_for
from .wiki import (
    WikiClient,
    article_url,
    extract_tag_region,
    from_url_or_title,
    slug,
    split_paragraphs,
    strip_wikitext,
    word_count,
)


async def _fetch_one(client: WikiClient, title: str) -> dict:
    """Fetch + extract; returns a record missing only scores."""
    record = {"title": title, "url": article_url(title)}
    try:
        wt, revid = await client.revision_of(title)
        if not wt or wt.lstrip().lower().startswith("#redirect"):
            record["skipped"] = "missing or redirect"
            return record
        region, whole = extract_tag_region(wt, "ai")
        text = strip_wikitext(region)[: config.SWEEP_MAX_CHARS]
        if len(text) < config.SWEEP_MIN_CHARS:
            record["skipped"] = f"too little prose ({len(text)} chars)"
            return record
        record.update(
            revid=revid,
            region="whole article" if whole else "tagged section",
            chars=len(text),
            text=text,
        )
    except Exception as e:
        record["error"] = str(e)
    return record


def _score_record(record: dict, scorer, cache: ScoreCache, calibration) -> dict:
    text = record["text"]
    paras = [p for p in split_paragraphs(text) if word_count(p) >= config.MIN_PARA_WORDS]
    paras = paras[: config.MAX_PARAS] if len(paras) > 1 else []

    texts = [text, *paras]
    ps: list[float | None] = [cache.get(scorer.model_id, t) for t in texts]
    todo = [i for i, p in enumerate(ps) if p is None]
    if todo:
        fresh = scorer.score_batch([texts[i] for i in todo])
        for i, p in zip(todo, fresh):
            ps[i] = round(p, 4)
        cache.put_many(scorer.model_id, [(texts[i], ps[i]) for i in todo])

    doc_p, *para_ps = ps
    record["pAI"] = doc_p
    record["verdict"] = verdict_for(doc_p, calibration)["verdict"]
    record["paragraphs"] = sorted(
        ({"p": p, "text": t} for t, p in zip(paras, para_ps)),
        key=lambda x: -x["p"],
    )
    return record


def _write_report(path, cat_title: str, model: str, calibration, done: dict, total: int):
    results = sorted(
        (r for r in done.values() if r.get("pAI") is not None),
        key=lambda r: -r["pAI"],
    )
    skipped = [r for r in done.values() if r.get("skipped") or r.get("error")]
    buckets = {
        "AI-typical": sum(r["verdict"] == "AI-typical" for r in results),
        "leaning AI": sum(r["verdict"].startswith("leaning") for r in results),
        "human-typical": sum(r["verdict"] == "human-typical" for r in results),
    }
    lines = [f"# AI-text sweep: {cat_title.removeprefix('Category:')}", ""]
    lines.append(
        f"Detector: {model}"
        + (
            f" (AUROC {calibration['auroc']} on {calibration['samples']['ai']} AI / {calibration['samples']['human']} human benchmark samples)"
            if calibration
            else " (uncalibrated)"
        )
    )
    lines.append(
        f"Scored {len(results)} of {total} articles ({len(skipped)} skipped/errored). "
        f"Verdicts: {', '.join(f'{v} {k}' for k, v in buckets.items())}."
    )
    lines += ["", "Detector output is triage, not proof — verify against sources and history before acting.", ""]
    lines += ["| p(AI) | verdict | article | worst paragraph |", "|---|---|---|---|"]
    for r in results:
        worst = r["paragraphs"][0] if r.get("paragraphs") else None
        preview = ""
        if worst:
            t = worst["text"]
            preview = f"{worst['p']:.3f} — {(t[:97] + '...' if len(t) > 100 else t)}".replace("|", "\\|")
        lines.append(f"| {r['pAI']:.3f} | {r['verdict']} | [{r['title'].replace('|', chr(92) + '|')}]({r['url']}) | {preview} |")
    if skipped:
        lines += ["", "## Skipped", ""]
        lines += [f"- [{r['title']}]({r['url']}) — {r.get('skipped') or r.get('error')}" for r in skipped]
    path.write_text("\n".join(lines) + "\n")


async def sweep(category: str, limit: int = 100000, out: str | None = None, model: str = config.DEFAULT_MODEL):
    from .models import Scorer

    cat_title = from_url_or_title(category)
    config.SWEEPS.mkdir(parents=True, exist_ok=True)
    base = config.SWEEPS / (out or slug(cat_title.removeprefix("Category:")))
    jsonl_path = base.with_suffix(".jsonl")
    report_path = base.with_suffix(".md")

    done: dict[str, dict] = {}
    if jsonl_path.exists():
        for line in jsonl_path.read_text().splitlines():
            if line.strip():
                r = json.loads(line)
                done[r["title"]] = r
        print(f"resuming: {len(done)} article(s) already in {jsonl_path}")

    client = WikiClient()
    print(f"listing {cat_title}...")
    titles = await client.category_members(cat_title)
    print(f"{len(titles)} member articles\n")

    calibration = load_calibration(model)
    if not calibration:
        print("⚠ no calibration for the detector — run: wikidetect calibrate\n")
    print(f"loading {model}...")
    scorer = Scorer(model)
    cache = ScoreCache()

    todo = [t for t in titles if t not in done]
    scored = sum(1 for r in done.values() if r.get("pAI") is not None)
    errors = 0

    # Fetching (network-bound) overlaps scoring (GPU-bound): fetch tasks fill
    # a queue that the scoring loop drains in arrival order.
    fetch_tasks = asyncio.Queue(maxsize=config.FETCH_CONCURRENCY * 4)

    async def producer():
        async def one(title):
            await fetch_tasks.put(await _fetch_one(client, title))

        pending = set()
        for title in todo:
            if scored >= limit:
                break
            pending.add(asyncio.create_task(one(title)))
            while len(pending) >= config.FETCH_CONCURRENCY:
                _, pending = await asyncio.wait(pending, return_when=asyncio.FIRST_COMPLETED)
        if pending:
            await asyncio.wait(pending)
        await fetch_tasks.put(None)  # sentinel

    prod = asyncio.create_task(producer())
    with jsonl_path.open("a") as jsonl:
        i = len(done)
        while True:
            record = await fetch_tasks.get()
            if record is None:
                break
            i += 1
            if scored >= limit:
                continue
            if "text" in record:
                # GPU work off the event loop so fetches keep flowing
                record = await asyncio.to_thread(_score_record, record, scorer, cache, calibration)
                scored += 1
                errors = 0
            elif record.get("error"):
                errors += 1
                if errors > 20:
                    print("too many consecutive errors — aborting (rerun to resume)")
                    break
            jsonl.write(json.dumps(record, ensure_ascii=False) + "\n")
            jsonl.flush()
            done[record["title"]] = record
            tag = f"{record['pAI']:.3f}" if record.get("pAI") is not None else ("skip " if record.get("skipped") else "ERR  ")
            print(f"[{i}/{len(titles)}] {tag}  {record['title']}")
    prod.cancel()
    await client.aclose()

    _write_report(report_path, cat_title, model, calibration, done, len(titles))
    results = sum(1 for r in done.values() if r.get("pAI") is not None)
    print(f"\nscored {results}/{len(titles)} ({len(done) - results} skipped)")
    print(f"raw results : {jsonl_path}")
    print(f"report      : {report_path}")

File: detector/sweep.js
#!/usr/bin/env node
// Overnight triage sweep: score EVERY article in a WikiProject AI Cleanup
// category with the benchmarked detector and write a ranked report.
//
//   npm run sweep -- --category="https://en.wikipedia.org/wiki/Category:Articles_containing_suspected_AI-generated_texts_from_November_2025"
//
//   # keep the Mac awake for an unattended run:
//   caffeinate -i npm run sweep -- --category="Category:..."
//
// Options:
//   --limit=N     stop after N scored articles (default: whole category)
//   --out=path    output basename (default: detector/sweeps/<category-slug>)
//
// Resumable by design: results append to <out>.jsonl as each article
// finishes; rerunning the same command skips everything already scored
// and rewrites the ranked <out>.md report at the end. Wikipedia calls are
// rate-limited (400ms + backoff) and the model loads once for the run.
//
// Scores each article's tag-covered region (whole article or tagged
// section), doc-level plus per-paragraph, so the report both ranks
// articles and points at the worst paragraphs.

import { appendFileSync, writeFileSync, readFileSync, mkdirSync, existsSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { fromUrlOrTitle, categoryMembers, wikitextOf, stripWikitext, extractTagRegion, slug, articleUrl } from './wiki.js';
import { startScorer, loadCalibration, verdictFor } from './scorer.js';

const MODEL = 'desklib';
const MIN_CHARS = 400;      // skip stubs — too little prose to score
const MIN_PARA_WORDS = 25;  // paragraph scores below this are noise
const MAX_PARAS = 10;

const argv = process.argv.slice(2);
const opt = (name, dflt = null) => {
  const arg = argv.find((a) => a.startsWith(`--${name}=`));
  return arg ? arg.split('=').slice(1).join('=') : dflt;
};
const category = opt('category');
const limit = parseInt(opt('limit', '100000'), 10);
if (!category) {
  console.error('usage: node detector/sweep.js --category="Category:... or URL" [--limit=N] [--out=path]');
  process.exit(1);
}

const catTitle = fromUrlOrTitle(category);
const sweepsDir = fileURLToPath(new URL('sweeps/', import.meta.url));
mkdirSync(sweepsDir, { recursive: true });
const outBase = opt('out', `${sweepsDir}${slug(catTitle.replace(/^Category:/i, ''))}`);
const jsonlPath = `${outBase}.jsonl`;
const reportPath = `${outBase}.md`;

// resume: everything already in the jsonl is done
const done = new Map();
if (existsSync(jsonlPath)) {
  for (const line of readFileSync(jsonlPath, 'utf8').split('\n').filter(Boolean)) {
    const r = JSON.parse(line);
    done.set(r.title, r);
  }
  console.log(`resuming: ${done.size} article(s) already in ${jsonlPath}`);
}

console.log(`listing ${catTitle}...`);
const titles = await categoryMembers(catTitle);
console.log(`${titles.length} member articles\n`);

const calibration = loadCalibration(MODEL);
if (!calibration) console.log('⚠ no calibration for the detector — run: npm run calibrate\n');
const scorer = startScorer(MODEL);

let scored = 0;
let errors = 0;
for (const [idx, title] of titles.entries()) {
  if (scored >= limit) break;
  if (done.has(title)) continue;

  let record = { title, url: articleUrl(title) };
  try {
    const wt = await wikitextOf(title);
    if (!wt || /^\s*#redirect/i.test(wt)) {
      record.skipped = 'missing or redirect';
    } else {
      const [region, wholeArticle] = extractTagRegion(wt, 'ai');
      const text = stripWikitext(region).slice(0, 12000);
      if (text.length < MIN_CHARS) {
        record.skipped = `too little prose (${text.length} chars)`;
      } else {
        const pAI = await scorer.score(text);
        const paras = text.split(/\n\s*\n/).map((p) => p.trim())
          .filter((p) => p.split(/\s+/).length >= MIN_PARA_WORDS)
          .slice(0, MAX_PARAS);
        const paraScores = [];
        for (const p of paras.length > 1 ? paras : []) {
          paraScores.push({
            p: await scorer.score(p),
            preview: p.length > 100 ? p.slice(0, 97) + '...' : p,
          });
        }
        record = {
          ...record,
          pAI,
          verdict: verdictFor(pAI, calibration).verdict,
          region: wholeArticle ? 'whole article' : 'tagged section',
          chars: text.length,
          paragraphs: paraScores.sort((a, b) => b.p - a.p),
        };
        scored++;
      }
    }
  } catch (e) {
    record.error = e.message;
    errors++;
    if (errors > 20) {
      console.error('too many consecutive errors — aborting (rerun to resume)');
      break;
    }
  }
  if (!record.error) errors = 0;
  appendFileSync(jsonlPath, JSON.stringify(record) + '\n');
  done.set(title, record);
  const tag = record.pAI !== undefined ? record.pAI.toFixed(3) : (record.skipped ? 'skip ' : 'ERR  ');
  console.log(`[${idx + 1}/${titles.length}] ${tag}  ${title}`);
}
scorer.close();

// ranked report
const results = [...done.values()].filter((r) => r.pAI !== undefined).sort((a, b) => b.pAI - a.pAI);
const skipped = [...done.values()].filter((r) => r.skipped || r.error);
const buckets = {
  'AI-typical': results.filter((r) => r.verdict === 'AI-typical').length,
  'leaning AI': results.filter((r) => r.verdict?.startsWith('leaning')).length,
  'human-typical': results.filter((r) => r.verdict === 'human-typical').length,
};

const lines = [];
lines.push(`# AI-text sweep: ${catTitle.replace(/^Category:/i, '')}`);
lines.push('');
lines.push(`Detector: ${MODEL}${calibration ? ` (AUROC ${calibration.auroc} on ${calibration.samples.ai} AI / ${calibration.samples.human} human benchmark samples)` : ' (uncalibrated)'}`);
lines.push(`Scored ${results.length} of ${titles.length} articles (${skipped.length} skipped/errored). Verdicts: ${Object.entries(buckets).map(([k, v]) => `${v} ${k}`).join(', ')}.`);
lines.push('');
lines.push('Detector output is triage, not proof — verify against sources and history before acting.');
lines.push('');
lines.push('| p(AI) | verdict | article | worst paragraph |');
lines.push('|---|---|---|---|');
for (const r of results) {
  const worst = r.paragraphs?.[0];
  const cell = worst ? `${worst.p.toFixed(3)} — ${worst.preview.replaceAll('|', '\\|')}` : '';
  lines.push(`| ${r.pAI.toFixed(3)} | ${r.verdict} | [${r.title.replaceAll('|', '\\|')}](${r.url}) | ${cell} |`);
}
if (skipped.length) {
  lines.push('');
  lines.push('## Skipped');
  lines.push('');
  for (const r of skipped) lines.push(`- [${r.title}](${r.url}) — ${r.skipped ?? r.error}`);
}
writeFileSync(reportPath, lines.join('\n') + '\n');

console.log(`\nscored ${results.length}/${titles.length} (${skipped.length} skipped)`);
console.log(`raw results : ${jsonlPath}`);
console.log(`report      : ${reportPath}`);
